mlflow/utils/databricks_utils.py

```diff
@@ -256,11 +258,102 @@ def is_in_databricks_runtime():
     return get_databricks_runtime_version() is not None
 
 
-def is_in_databricks_serverless():
+def is_in_databricks_serverless_runtime():
     dbr_version = get_databricks_runtime_version()
     return dbr_version and dbr_version.startswith("client.")
 
 
+def is_in_databricks_shared_cluster_runtime():
+    from mlflow.utils.spark_utils import is_spark_connect_mode
+
+    return (
+        is_in_databricks_runtime()
+        and is_spark_connect_mode()
+        and not is_in_databricks_serverless_runtime()
+    )
+
+
+def is_databricks_connect(spark):
+    """
+    Return True if current Spark-connect client connects to Databricks cluster.
+    """
+    from mlflow.utils.spark_utils import is_spark_connect_mode
+
+    if not is_spark_connect_mode():
+        return False
+
+    if is_in_databricks_serverless_runtime() or is_in_databricks_shared_cluster_runtime():
+        return True
+    try:
+        # TODO: Remove the `spark.client._builder._build` attribute access once
+        #  Spark-connect has public attribute for this information.
+        return "databricks-session" in spark.client._builder.userAgent
+    except Exception:
+        return False
+
+
+@dataclass
+class DBConnectClientCache:
+    client: "pyspark.sql.connect.session.SparkSession"  # noqa: F821
```

Comment:
You need to import  

```
if TYPE_CHECKING:
    from pyspark.sql.connect.session import SparkSession
```

should fix the error

Reply:
I don't think we should do this, use might use old version pyspark which doesn't have spark-connect class.

Reply:
The code in `if TYPE_CHECKING` isn't executed at runtime. See https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Reply:
ok make sense.